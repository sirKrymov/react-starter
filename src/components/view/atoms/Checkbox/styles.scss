/* stylelint-disable selector-max-compound-selectors, max-nesting-depth */
@import '../../../../styles/configs';
@import '../../../../styles/mixins';

.checkbox {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  position: relative;
  display: inline-block;
  width: 16px;

  &--disabled {
    pointer-events: none;
    opacity: 0.6;
    cursor: not-allowed;
  }

  input {
    display: none;

    &:checked + .checkbox__label {
      background-color: var(--valid);

      .checkbox__icon {
        opacity: 1;
      }
    }
  }

  &__label {
    background-color: var(--borderDark);
    border-radius: 2px;
    transition: background-color $anim-time-fast;
    position: relative;
    overflow: hidden;
    display: block;
    cursor: pointer;
    height: 16px;
    width: 100%;
  }

  &__icon {
    transition: opacity $anim-time-fast;
    transform: translate(-50%, -50%);
    font-size: 20px;
    position: absolute;
    opacity: 0;
    height: 18px;
    color: var(--primary);
    width: 18px;
    left: 50%;
    top: 50%;
    @include center;

    svg {
      vertical-align: initial;
    }
  }
}
