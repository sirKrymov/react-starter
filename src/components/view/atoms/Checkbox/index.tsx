// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ChangeEvent, ReactElement } from 'react';

import { IoMdCheckmark } from 'react-icons/io';
import cn from 'classnames';

import './styles.scss';

interface IProps {
  defaultChecked?: boolean;
  onChange(e: ChangeEvent<any>): void;
  className?: string;
  disabled?: boolean;
  value: boolean;
  size?: string;
  name: string;
  id: string;
}

export function Checkbox({
  defaultChecked,
  className,
  onChange,
  disabled,
  value,
  size,
  name,
  id
}: IProps): ReactElement {
  const checkboxClass = cn(
    {
      checkbox: true,
      [`checkbox--size-${size}`]: size,
      'checkbox--disabled': disabled
    },
    className
  );

  return (
    <div className={checkboxClass}>
      <input
        defaultChecked={defaultChecked}
        onChange={onChange}
        disabled={disabled}
        checked={value}
        type="checkbox"
        name={name}
        id={id}
      />

      <label className="checkbox__label" htmlFor={id}>
        <span className="checkbox__icon">
          <IoMdCheckmark />
        </span>
      </label>
    </div>
  );
}
