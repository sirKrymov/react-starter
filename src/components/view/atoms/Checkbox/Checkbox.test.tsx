import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';

import { Checkbox } from '../Checkbox';

describe('<Switch> atom component', () => {
  const handleChange = sinon.spy();
  const handleClick = sinon.spy();
  const chkID = 'checkboxID';

  let shallowComponent: ShallowWrapper;

  const createComponent = (props?: any) => {
    shallowComponent = shallow(
      <Checkbox
        id={chkID}
        name="checkbox"
        value={false}
        onChange={handleChange}
        {...props}
      />
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent();
    });

    it('THEN it should render without crashing', () => {
      expect(shallowComponent.html()).not.toHaveLength(0);
    });

    it('THEN it should match snapshot', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot();
    });
  });

  describe('WHEN the component have disabled status', () => {
    it('AND it should disable switch slider if it contains disabled props', () => {
      createComponent({ disabled: true });

      expect(shallowComponent.find('#' + chkID).props().disabled).toBe(true);
      expect(shallowComponent.find('div.checkbox--disabled')).toHaveLength(1);
      shallowComponent.find('#' + chkID).simulate('click');
      handleClick.calledOnceWith(undefined);
    });
  });
});
