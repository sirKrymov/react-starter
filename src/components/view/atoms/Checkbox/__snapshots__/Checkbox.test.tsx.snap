// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch> atom component WHEN the component loads THEN it should match snapshot 1`] = `
<div
  className="checkbox"
>
  <input
    checked={false}
    id="checkboxID"
    name="checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="checkbox__label"
    htmlFor="checkboxID"
  >
    <span
      className="checkbox__icon"
    >
      <IoMdCheckmark />
    </span>
  </label>
</div>
`;
